internal/resource/tree.go

```diff
@@ -0,0 +1,164 @@
+package resource
+
+import (
+	"encoding/json"
+	"slices"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/emirpasic/gods/v2/trees/redblacktree"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type indexedResource struct {
+	Resource            *Resource
+	State               *apiv1.ResourceState
+	PendingDependencies map[Ref]struct{}
+	Dependents          map[Ref]*indexedResource
+}
+
+// treeBuilder is used to index a set of resources into a stateTree.
+type treeBuilder struct {
+	byRef        map[Ref]*indexedResource
+	byGroup      *redblacktree.Tree[int, []*indexedResource]
+	byDefiningGK map[schema.GroupKind]*indexedResource
+	byGK         map[schema.GroupKind]*indexedResource
+}
+
+func (b *treeBuilder) Add(resource *Resource) {
+	// Initialize the builder
+	if b.byRef == nil {
+		b.byRef = map[Ref]*indexedResource{}
+	}
+	if b.byGroup == nil {
+		b.byGroup = redblacktree.New[int, []*indexedResource]()
+	}
+	if b.byDefiningGK == nil {
+		b.byDefiningGK = map[schema.GroupKind]*indexedResource{}
+	}
+	if b.byGK == nil {
+		b.byGK = map[schema.GroupKind]*indexedResource{}
+	}
+
+	// Handle conflicting refs deterministically
+	if existing, ok := b.byRef[resource.Ref]; ok && resource.Less(existing.Resource) {
```

Comment:
Should we be referencing a field within the manifest rather than the whole manifest? Or does it not matter which one wins for the purposes of this tree?

Reply:
It doesn't matter which wins as long as the same one will win every time regardless of order